knative-operator/pkg/webhook/knativekafka/webhook_validating.go

```diff
@@ -0,0 +1,139 @@
+	"sigs.k8s.io/controller-runtime/pkg/webhook"
+	"sigs.k8s.io/controller-runtime/pkg/webhook/admission"
+	"sigs.k8s.io/controller-runtime/pkg/webhook/admission/builder"
+	"sigs.k8s.io/controller-runtime/pkg/webhook/admission/types"
+)
+
+// Creates a new validating KnativeKafka Webhook
+func ValidatingWebhook(mgr manager.Manager) (webhook.Webhook, error) {
+	common.Log.Info("Setting up validating webhook for KnativeKafka")
+	return builder.NewWebhookBuilder().
+		Name("validating.knativekafka.openshift.io").
+		Validating().
+		Operations(admissionregistrationv1beta1.Create, admissionregistrationv1beta1.Update).
+		WithManager(mgr).
+		ForType(&operatorv1alpha1.KnativeKafka{}).
+		Handlers(&KnativeKafkaValidator{}).
+		Build()
+}
+
+// KnativeKafkaValidator validates KnativeKafka CR's
+type KnativeKafkaValidator struct {
+	client  client.Client
+	decoder types.Decoder
+}
+
+// Implement admission.Handler so the controller can handle admission request.
+var _ admission.Handler = (*KnativeKafkaValidator)(nil)
+
+// What makes us a webhook
+func (v *KnativeKafkaValidator) Handle(ctx context.Context, req types.Request) types.Response {
+	ke := &operatorv1alpha1.KnativeKafka{}
+
+	err := v.decoder.Decode(req, ke)
+	if err != nil {
+		return admission.ErrorResponse(http.StatusBadRequest, err)
+	}
+
+	allowed, reason, err := v.validate(ctx, ke)
+	if err != nil {
+		return admission.ErrorResponse(http.StatusInternalServerError, err)
+	}
+	return admission.ValidationResponse(allowed, reason)
+}
+
+// KnativeKafkaValidator checks for a minimum OpenShift version
+func (v *KnativeKafkaValidator) validate(ctx context.Context, ke *operatorv1alpha1.KnativeKafka) (allowed bool, reason string, err error) {
+	log := common.Log.WithName("validate")
+	stages := []func(context.Context, *operatorv1alpha1.KnativeKafka) (bool, string, error){
+		v.validateNamespace,
+		v.validateVersion,
+		v.validateLoneliness,
+		v.validateShape,
+	}
+	for _, stage := range stages {
+		allowed, reason, err = stage(ctx, ke)
+		if len(reason) > 0 {
+			if err != nil {
+				log.Error(err, reason)
+			} else {
+				log.Info(reason)
+			}
+		}
+		if !allowed {
+			return
+		}
+	}
+	return
+}
+
+// KnativeKafkaValidator implements inject.Client.
+// A client will be automatically injected.
+var _ inject.Client = (*KnativeKafkaValidator)(nil)
+
+// InjectClient injects the client.
+func (v *KnativeKafkaValidator) InjectClient(c client.Client) error {
+	v.client = c
+	return nil
+}
+
+// KnativeKafkaValidator implements inject.Decoder.
+// A decoder will be automatically injected.
+var _ inject.Decoder = (*KnativeKafkaValidator)(nil)
+
+// InjectDecoder injects the decoder.
+func (v *KnativeKafkaValidator) InjectDecoder(d types.Decoder) error {
+	v.decoder = d
+	return nil
+}
+
+// validate minimum openshift version
+func (v *KnativeKafkaValidator) validateVersion(ctx context.Context, _ *operatorv1alpha1.KnativeKafka) (bool, string, error) {
```

Comment:
Do we even need to do this in this webhook? Seems unnecessary

Reply:
Do you mean that because `KnativeEventing` is a prerequisite and we already check the OpenShift version there?

Reply:
Meh, seems like we'd do this in a different place anyway. I'd actually expect Serverless to never become ready and I think we should revisit this check at all since OLM now has mechanisms for this too.

Do you mind filing a JIRA to revisit the version check in webhooks and drop it from this specific webhook?

Reply:
Sounds good, I will do it

Reply:
Created https://issues.redhat.com/browse/SRVCOM-1071